### PR TITLE
* Fix issue with info banner display getting cut off in gallery view

### LINKED ISSUE #5589: Notification banner gets cut off in Gallery List view
**What** & **Where**

Steps to reproduce 

1. Open in incognito to make sure cookies have not been set to view the notification banner
2. Search for something in catalogue that has lots of images - e.g hhh
3. Click on gallery view  -> https://wellcomecollection.org/works/x9wbqn69/items?sierraId=b28738512&langCode=eng&canvas=1
4. Click on the view all icon to collapse the preview of image. 

Result:

* Notification banner is hidden 
* Banner has padding issue
* Viewed on Chrome
* Page is still functional and is more a cosmetic issue. 

Desktop 
Screenshot 2020-10-02 at 16 56 52

Mobile 

Screenshot 2020-10-02 at 17 04 58

### PR DESCRIPTION
**Pull request by ishouty**
* Fix issue with info banner display getting cut off in gallery view #5728
* Fix issue Flickering screen when displaying the scroll bar on the works viewer #5726

* Implemented context provider to be able to update state for GlobalAlerts infobanner, (ability to update global state - when required) to make sure we can read the offset of the topbar and to trigger update view on gridviewer.
* added overflow hidden and scrollTo to GridViewer to stop the flickering issue and scrolling which is not needed
* Update infoBanner from js to typescript

Testing 
I am happy to add some unit tests but I would like to make sure you guys are happy with the approach.

### CODE REVIEW

File: common/views/components/IIIFViewer/parts/GridViewer.js [hunk 1]
```diff
@@ -103,27 +109,48 @@ const GridViewer = ({
...
-
   useEffect(() => {
     const rowIndex = Math.floor(activeIndex / columnCount);
     grid &&
       grid.current &&
       grid.current.scrollToItem({ align: 'start', rowIndex });
   }, [activeIndex]);
+  useEffect(() => {
+    setTimeout(() => {
+      window.scrollTo(0, 0);
```
Comment (ishouty) [inline common/views/components/IIIFViewer/parts/GridViewer.js:129]: this part is very adhoc, but its difficult to do this other ways.

File: common/views/components/IIIFViewer/parts/GridViewer.js [hunk 2]
```diff
@@ -103,27 +109,48 @@ const GridViewer = ({
...
+    }, 500);
+
+    const body = document && document.body ? document.body : null;
+    // there are multiple scrolls in this view, we have to set the body to hidden to stop flickering and offset
+    if (body && body.style) {
+      body.style.overflow = 'hidden';
```
Comment (ishouty) [inline common/views/components/IIIFViewer/parts/GridViewer.js:135]: gridviewer at times feels like it needs to know more about state management of other components.

File: common/views/components/GlobalAlertContext/GlobalAlertContext.tsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    });
+  };
+
+  // initial context value
+  state = {
+    infoBanner: false,
+    isShown: this.props.value.isShown,
+    routeRegex: this.props.value.routeRegex,
+    text: this.props.value.text,
+    setInfoBanner: this.setInfoBanner,
```
Comment (ishouty) [inline common/views/components/GlobalAlertContext/GlobalAlertContext.tsx:36]: children components can set state which can be passed around to update state  to other components.

Comment (jamesgorrie) [reply common/views/components/GlobalAlertContext/GlobalAlertContext.tsx:36]: Should we convert this to hooks and a functional component while we're at it?

File: common/views/components/IIIFViewer/parts/GridViewer.js (hunk 1 above)
Comment (jamesgorrie) [reply common/views/components/IIIFViewer/parts/GridViewer.js:129]: Could you leave a comment as to why this is like this, maybe a link to an example? I can't tell from just looking at it, and I might remove it as a future develop unless I knew why it was there.

File: common/views/components/IIIFViewer/parts/GridViewer.js (hunk 2 above)
Comment (jamesgorrie) [reply common/views/components/IIIFViewer/parts/GridViewer.js:135]: Could you use `document?.body.?style.overflow = 'hidden`?

File: common/views/components/GlobalAlertContext/GlobalAlertContext.tsx [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    //Passing the state object as a value prop to all children
+    return (
+      <GlobalAlertContext.Provider value={this.state}>
+        {this.props.children}
+      </GlobalAlertContext.Provider>
+    );
+  }
+}
+
+export default GlobalAlertContext;
```
Comment (jamesgorrie) [inline common/views/components/GlobalAlertContext/GlobalAlertContext.tsx:49]: Could we call this either `GlobalAlert`, or `GlobalAlertContextProvider` so we don't sonfure it with the standard implementation of a `Context`?

**Review by jamesgorrie — COMMENTED** [review]
I'm wondering if we have a `GlobalAlertContext` where we can use

```TS
const { isVisible: globalContextIsVisible } = useContext(GlobalAlertContext);
```

So we can use use it with the standard APIs?

File: common/views/components/IIIFViewer/parts/GridViewer.js (hunk 2 above)
Comment (ishouty) [reply common/views/components/IIIFViewer/parts/GridViewer.js:135]: I tried that but flow didnt like it .. i could add flowfixme and add the other way in.

File: common/views/components/IIIFViewer/parts/GridViewer.js (hunk 1 above)
Comment (ishouty) [reply common/views/components/IIIFViewer/parts/GridViewer.js:129]: yep, will do

File: common/views/components/GlobalAlertContext/GlobalAlertContext.tsx (hunk 4 above)
Comment (ishouty) [reply common/views/components/GlobalAlertContext/GlobalAlertContext.tsx:49]: yep GlobalAlert sounds good.

File: common/views/components/GlobalAlertContext/GlobalAlertContext.tsx (hunk 3 above)
Comment (ishouty) [reply common/views/components/GlobalAlertContext/GlobalAlertContext.tsx:36]: Do you mean converting this existing file to a functional component ? and using hooks to save set of setInfoBanner?

**Comment by ishouty** [comment]
> I'm wondering if we have a `GlobalAlertContext` where we can use
> 
> ```ts
> const { isVisible: globalContextIsVisible } = useContext(GlobalAlertContext);
> ```
> 
> So we can use use it with the standard APIs?

do you mean renaming the setInfoBanner to a better standard .. isVisible ?

File: common/views/components/IIIFViewer/parts/GridViewer.js (hunk 2 above)
Comment (jamesgorrie) [reply common/views/components/IIIFViewer/parts/GridViewer.js:136]: Na, we should adhere to the types as that's where we get safety. What was the error?

**Comment by jamesgorrie** [comment]
> do you mean renaming the setInfoBanner to a better standard .. isVisible ?

I was wondering if we just use a standard Context so that the API for that context uses the above rather than using the old state APIs.

File: common/views/components/IIIFViewer/parts/GridViewer.js (hunk 2 above)
Comment (ishouty) [reply common/views/components/IIIFViewer/parts/GridViewer.js:136]: document?.body?.style?.overflow = 'hidden';

Invalid left-hand side in assignment

File: common/views/components/InfoBanner/InfoBanner.tsx [hunk 5]
```diff
@@ -1,85 +1,94 @@
...
-type State = {
-  showInfoBanner: boolean;
 };
 
 const InfoBanner: FunctionComponent<Props> = ({ cookieName, text }: Props) => {
-  const [isVisible, setIsVisible] = useState<boolean>(false);
+  const { showInfoBarBanner, setInfoBarBanner } = useContext(
+    GlobalInfoBarContext
+  );
+  useEffect(() => {
```
Comment (jamesgorrie) [inline common/views/components/InfoBanner/InfoBanner.tsx:24]: it looks like we've pulled in some changes that were made [here](https://github.com/wellcomecollection/wellcomecollection.org/commit/c023b81d4a81f56f10080cfb9cb608e040a7789a#diff-6d893f831dfaa9863006df4faf1b11600e85441c6463a594717d543f2e449183).

**Pull request closed without merging**